libbpf-rs/tests/test_netfilter.rs

```diff
@@ -0,0 +1,124 @@
+#[allow(dead_code)]
+mod common;
+
+use std::process::Command;
+
+use libbpf_rs::Map;
+use libbpf_rs::NetfilterOpts;
+use libbpf_rs::Object;
+use test_tag::tag;
+
+use libbpf_rs::NFPROTO_IPV4;
+use libbpf_rs::NFPROTO_IPV6;
+
+// use libbpf_rs::NF_INET_FORWARD;
+use libbpf_rs::NF_INET_LOCAL_IN;
+// use libbpf_rs::NF_INET_LOCAL_OUT;
+use libbpf_rs::NF_INET_POST_ROUTING;
+use libbpf_rs::NF_INET_PRE_ROUTING;
+
+use crate::common::bump_rlimit_mlock;
+use crate::common::get_map_mut;
+use crate::common::get_prog_mut;
+use crate::common::get_test_object;
+
+// copy from test
+// I wasn't sure if I could just move this function to common so I copied it
+fn with_ringbuffer<F>(map: &Map, action: F) -> i32
+where
+    F: FnOnce(),
+{
+    let mut value = 0i32;
+    {
+        let callback = |data: &[u8]| {
+            plain::copy_from_bytes(&mut value, data).expect("Wrong size");
+            0
+        };
+
+        let mut builder = libbpf_rs::RingBufferBuilder::new();
+        builder.add(map, callback).expect("failed to add ringbuf");
+        let mgr = builder.build().expect("failed to build");
+
+        action();
+        mgr.consume().expect("failed to consume ringbuf");
+    }
+
+    value
+}
+
+fn test_attach_and_detach(obj: &mut Object, pf: u32, hooknum: u32, hook_desc: &str) {
+    let prog = get_prog_mut(obj, "handle_netfilter");
+    let netfilter_opt = libbpf_rs::NetfilterOpts {
+        pf,
+        hooknum,
+        ..NetfilterOpts::default()
+    };
+    let error_message = format!(
+        "Failed to attach netfilter protocol {}, hook: {}",
+        pf, hook_desc
+    );
+    let link = prog.attach_netfilter(netfilter_opt).expect(&error_message);
+
+    let map = get_map_mut(obj, "ringbuf");
+
+    if pf == 2 {
+        let result = match hook_desc {
+            "PRE_ROUTING" | "LOCAL_IN" | "LOCAL_OUT" | "POST_ROUTING" => {
+                let action = || {
+                    Command::new("sh")
+                        .arg("-c")
+                        .arg("echo 'Test data' | nc -u 127.0.0.1 12345")
+                        .output()
+                        .expect("Failed to send packet");
```

Comment:
Let's please not depend on `nc` being installed. Can this just be expressed using, say, https://doc.rust-lang.org/std/net/struct.TcpStream.html instead?